File: taskhub/lib/models/professional.dart
import 'package:taskhub/models/citizen.dart';

class Education {
  final String? institution;
  final String? degree;
  final String? fieldOfStudy;
  final int? startYear;
  final int? endYear;
  final String? description;

  Education({
    this.institution,
    this.degree,
    this.fieldOfStudy,
    this.startYear,
    this.endYear,
    this.description,
  });

  factory Education.fromJson(Map<String, dynamic> json) {
    return Education(
      institution: json['institution'],
      degree: json['degree'],
      fieldOfStudy: json['field_of_study'],
      startYear: json['start_year'],
      endYear: json['end_year'],
      description: json['description'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'institution': institution,
      'degree': degree,
      'field_of_study': fieldOfStudy,
      'start_year': startYear,
      'end_year': endYear,
      'description': description,
    };
  }
}

class Certification {
  final String? name;
  final String? issuer;
  final int? issueYear;
  final String? credentialUrl;

  Certification({this.name, this.issuer, this.issueYear, this.credentialUrl});

  factory Certification.fromJson(Map<String, dynamic> json) {
    return Certification(
      name: json['name'],
      issuer: json['issuer'],
      issueYear: json['issue_year'],
      credentialUrl: json['credential_url'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'name': name,
      'issuer': issuer,
      'issue_year': issueYear,
      'credential_url': credentialUrl,
    };
  }
}

class Professional {
  final int? id;
  final int userId;
  final String documentType;
  final String documentNumber;
  final String? documentIssuerState;
  final String? professionalRegistry;
  final String serviceType;
  final String? specialties;
  final String? description;
  final int yearsExperience;
  final String? portfolioUrl;
  final List<String>? portfolioItems;
  final List<Education>? education;
  final List<Certification>? certifications;
  final double? hourlyRate;
  final String availabilityStatus;
  final Map<String, dynamic>? workingHours;
  final Address? address;
  final double? rating;
  final int? totalReviews;

  Professional({
    this.id,
    required this.userId,
    required this.documentType,
    required this.documentNumber,
    this.documentIssuerState,
    this.professionalRegistry,
    required this.serviceType,
    this.specialties,
    this.description,
    this.yearsExperience = 0,
    this.portfolioUrl,
    this.portfolioItems,
    this.education,
    this.certifications,
    this.hourlyRate,
    this.availabilityStatus = 'available',
    this.workingHours,
    this.address,
    this.rating,
    this.totalReviews,
  });

  String get availabilityLabel {
    switch (availabilityStatus) {
      case 'available':
        return 'Disponível';
      case 'unavailable':
        return 'Indisponível';
      case 'on_vacation':
        return 'Em férias';
      default:
        return availabilityStatus;
    }
  }

  String get ratingDisplay {
    if (rating == null || rating == 0) return 'Sem avaliações';
    return '${rating!.toStringAsFixed(1)} ⭐ ($totalReviews avaliações)';
  }

  factory Professional.fromJson(Map<String, dynamic> json) {
    List<Education>? educationList;
    if (json['education'] != null) {
      educationList = (json['education'] as List)
          .map((e) => Education.fromJson(e))
          .toList();
    }

    List<Certification>? certificationsList;
    if (json['certifications'] != null) {
      certificationsList = (json['certifications'] as List)
          .map((c) => Certification.fromJson(c))
          .toList();
    }

    return Professional(
      id: json['id'],
      userId: json['user_id'] ?? 0,
      documentType: json['document_type'] ?? '',
      documentNumber: json['document_number'] ?? '',
      documentIssuerState: json['document_issuer_state'],
      professionalRegistry: json['professional_registry'],
      serviceType: json['service_type'] ?? '',
      specialties: json['specialties'],
      description: json['description'],
      yearsExperience: json['years_experience'] ?? 0,
      portfolioUrl: json['portfolio_url'],
      portfolioItems: json['portfolio_items'] != null
          ? List<String>.from(json['portfolio_items'])
          : null,
      education: educationList,
      certifications: certificationsList,
      hourlyRate: json['hourly_rate'] != null
          ? double.tryParse(json['hourly_rate'].toString())
          : null,
      availabilityStatus: json['availability_status'] ?? 'available',
      workingHours: json['working_hours'],
      address: json['address'] != null
          ? Address.fromJson(json['address'])
          : null,
      rating: json['rating'] != null
          ? double.tryParse(json['rating'].toString())
          : null,
      totalReviews: json['total_reviews'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'user_id': userId,
      'document_type': documentType,
      'document_number': documentNumber,
      'document_issuer_state': documentIssuerState,
      'professional_registry': professionalRegistry,
      'service_type': serviceType,
      'specialties': specialties,
      'description': description,
      'years_experience': yearsExperience,
      'portfolio_url': portfolioUrl,
      'portfolio_items': portfolioItems,
      'education': education?.map((e) => e.toJson()).toList(),
      'certifications': certifications?.map((c) => c.toJson()).toList(),
      'hourly_rate': hourlyRate,
      'availability_status': availabilityStatus,
      'working_hours': workingHours,
      'address': address?.toJson(),
    };
  }
}

File: backend/src/models/Professional.js
class Professional {
    constructor({
        id,
        user_id,
        document_type,
        document_number,
        document_issuer_state,
        professional_registry,
        service_type,
        specialties,
        description,
        years_experience,
        portfolio_url,
        portfolio_items,
        education,
        certifications,
        hourly_rate,
        availability_status,
        working_hours,
        address_street,
        address_number,
        address_complement,
        address_neighborhood,
        address_city,
        address_state,
        address_zip_code,
        address_country,
        rating,
        total_reviews,
        created_at,
        updated_at,
    }) {
        this.id = id;
        this.user_id = user_id;
        this.document_type = document_type;
        this.document_number = document_number;
        this.document_issuer_state = document_issuer_state;
        this.professional_registry = professional_registry;
        this.service_type = service_type;
        this.specialties = specialties;
        this.description = description;
        this.years_experience = years_experience;
        this.portfolio_url = portfolio_url;
        this.portfolio_items = typeof portfolio_items === 'string' ? JSON.parse(portfolio_items || '[]') : portfolio_items || [];
        this.education = typeof education === 'string' ? JSON.parse(education || '[]') : education || [];
        this.certifications = typeof certifications === 'string' ? JSON.parse(certifications || '[]') : certifications || [];
        this.hourly_rate = hourly_rate;
        this.availability_status = availability_status;
        this.working_hours = typeof working_hours === 'string' ? JSON.parse(working_hours || '{}') : working_hours || {};
        this.address_street = address_street;
        this.address_number = address_number;
        this.address_complement = address_complement;
        this.address_neighborhood = address_neighborhood;
        this.address_city = address_city;
        this.address_state = address_state;
        this.address_zip_code = address_zip_code;
        this.address_country = address_country;
        this.rating = rating;
        this.total_reviews = total_reviews;
        this.created_at = created_at;
        this.updated_at = updated_at;
    }

    /**
     * Get full address
     */
    getFullAddress() {
        const parts = [
            this.address_street,
            this.address_number,
            this.address_complement,
            this.address_neighborhood,
            this.address_city,
            `${this.address_state} ${this.address_zip_code}`,
            this.address_country,
        ].filter(Boolean);
        return parts.join(', ');
    }

    /**
     * Get availability label
     */
    getAvailabilityLabel() {
        const statuses = {
            available: 'Disponível',
            unavailable: 'Indisponível',
            on_vacation: 'Em férias',
        };
        return statuses[this.availability_status] || this.availability_status;
    }

    /**
     * Get rating display
     */
    getRatingDisplay() {
        if (!this.rating) return 'Sem avaliações';
        return `${this.rating.toFixed(1)} ⭐ (${this.total_reviews} avaliações)`;
    }

    /**
     * Convert to JSON for response
     */
    toJSON() {
        return {
            id: this.id,
            user_id: this.user_id,
            document_type: this.document_type,
            document_number: this.document_number,
            professional_registry: this.professional_registry,
            service_type: this.service_type,
            specialties: this.specialties,
            description: this.description,
            years_experience: this.years_experience,
            portfolio_url: this.portfolio_url,
            portfolio_items: this.portfolio_items,
            education: this.education,
            certifications: this.certifications,
            hourly_rate: this.hourly_rate,
            availability_status: this.availability_status,
            working_hours: this.working_hours,
            address: {
                street: this.address_street,
                number: this.address_number,
                complement: this.address_complement,
                neighborhood: this.address_neighborhood,
                city: this.address_city,
                state: this.address_state,
                zip_code: this.address_zip_code,
                country: this.address_country,
            },
            rating: this.rating,
            total_reviews: this.total_reviews,
        };
    }
}

module.exports = Professional;
